compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -273,55 +405,143 @@ pub fn push_debuginfo_type_name<'tcx>(
+            output.push_str("::");
+        }
+    }
+
+    push_unqualified_item_name(tcx, def_id, def_key.disambiguated_data, output);
+}
+
+fn push_unqualified_item_name(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    disambiguated_data: DisambiguatedDefPathData,
+    output: &mut String,
+) {
+    let cpp_like_names = tcx.sess.target.is_like_msvc;
 
-    fn push_item_name(tcx: TyCtxt<'tcx>, def_id: DefId, qualified: bool, output: &mut String) {
-        if qualified {
+    match disambiguated_data.data {
+        DefPathData::CrateRoot => {
             output.push_str(&tcx.crate_name(def_id.krate).as_str());
-            for path_element in tcx.def_path(def_id).data {
-                write!(output, "::{}", path_element.data).unwrap();
+        }
+        DefPathData::Impl => {
+            let self_ty = tcx.type_of(def_id);
+            output.push_str(if cpp_like_names { "impl$<" } else { "<impl " });
+            if let Some(impl_trait_ref) = tcx.impl_trait_ref(def_id) {
+                push_item_name(tcx, impl_trait_ref.def_id, true, output);
+                output.push_str(if cpp_like_names { ", " } else { " for " });
             }
-        } else {
-            output.push_str(&tcx.item_name(def_id).as_str());
+            push_debuginfo_type_name(tcx, self_ty, false, output, &mut FxHashSet::default());
+            push_close_angle_bracket(tcx, output);
         }
+        DefPathData::ClosureExpr if tcx.generator_kind(def_id).is_some() => {
+            // Generators look like closures, but we want to treat them differently
+            // in the debug info.
+            if cpp_like_names {
+                write!(output, "generator${}", disambiguated_data.disambiguator).unwrap();
+            } else {
+                write!(output, "{{generator#{}}}", disambiguated_data.disambiguator).unwrap();
+            }
+        }
+        _ => match disambiguated_data.data.name() {
+            DefPathDataName::Named(name) => {
+                output.push_str(&name.as_str());
+            }
+            DefPathDataName::Anon { namespace } => {
+                if cpp_like_names {
+                    write!(output, "{}${}", namespace, disambiguated_data.disambiguator).unwrap();
+                } else {
+                    write!(output, "{{{}#{}}}", namespace, disambiguated_data.disambiguator)
+                        .unwrap();
+                }
+            }
+        },
+    };
+}
+
+// Pushes the generic parameters in the given `InternalSubsts` to the output string.
+// This ignores region parameters, since they can't reliably be
+// reconstructed for items from non-local crates. For local crates, this
+// would be possible but with inlining and LTO we have to use the least
+// common denominator - otherwise we would run into conflicts.
+fn push_generic_params_internal<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    substs: SubstsRef<'tcx>,
+    normalize_type: bool,
```

Comment:
I don't think `normalize_type` should be a parameter. Instead the function should assume that `substs` is already normalized -- e.g. by adding something like `debug_assert!(substs == tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), substs))` at the beginning of the method. That should simplify the code.

Reply:
Fair enough